### PR TITLE
matrix/count_paths.py

### PR DESCRIPTION
**Pull request by jdevers1**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: matrix/count_paths.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+start here  ->   0  0  0  0
+                 1  1  0  0
+                 0  0  0  1
+                 0  1  0  0  <- finish here
+how many 'distinct' paths can you take to get to the finish?
+Using a recursive depth-first search algorithm below, you are able to
+find the number of distinct unique paths (count).
+
+'*' will demostrate a path
```
Comment (ChrisO345) [inline matrix/count_paths.py:13]:
```suggestion
'*' will demonstrate a path
```

**Review by ChrisO345 — CHANGES_REQUESTED** [review]

File: matrix/count_paths.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+In the example above, there are two distinct paths:
+1.                2.
+    *  *  *  0      *  *  *  *
+    1  1  *  0      1  1  *  *
+    0  0  *  1      0  0  *  1
+    0  1  *  *      0  1  *  *
+"""
+
+
+def dfs(grid: "list[list[int]]", row: int, col: int, visit: set) -> int:
```
Comment (CaedenPH) [inline matrix/count_paths.py:23]:
```suggestion
def depth_first_search(grid: list[list[int]], row: int, col: int, visit: set) -> int:
```

File: matrix/count_paths.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    0  0  0  0
+    1  1  0  0
+    0  0  0  1
+    0  1  0  0
+    >>> start_row = 0
+    >>> start_col = 0
+    >>> visited = set()
+    >>> grid = [[0, 0, 0, 0], [1, 1, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0]]
+    >>> dfs(grid, start_row, start_col, visited)
```
Comment (CaedenPH) [inline matrix/count_paths.py:40]:
```suggestion
    >>> grid = [[0, 0, 0, 0], [1, 1, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0]]
    >>> dfs(grid, 0, 0, set())
```

File: matrix/count_paths.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    0  0  0  0  0
+    0  1  1  1  0
+    0  1  1  1  0
+    0  0  0  0  0
+    >>> start_row = 0
+    >>> start_col = 0
+    >>> visited = set()
+    >>> grid = [[0, 0, 0, 0, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 0, 0, 0, 0]]
+    >>> dfs(grid, start_row, start_col, visited)
```
Comment (CaedenPH) [inline matrix/count_paths.py:51]:
```suggestion
    >>> grid = [[0, 0, 0, 0, 0], [0, 1, 1, 1, 0], [0, 1, 1, 1, 0], [0, 0, 0, 0, 0]]
    >>> dfs(grid, 0, 0, set())
```

**Review by CaedenPH — COMMENTED** [review]
Looks good - small changes

File: matrix/count_paths.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    if row == row_length - 1 and col == col_length - 1:
+        return 1
+
+    visit.add((row, col))
+
+    count = 0
+    count += dfs(grid, row + 1, col, visit)
+    count += dfs(grid, row - 1, col, visit)
+    count += dfs(grid, row, col + 1, visit)
+    count += dfs(grid, row, col - 1, visit)
```
Comment (ChrisO345) [inline matrix/count_paths.py:72]:
```suggestion
    count += depth_first_search(grid, row + 1, col, visit)
    count += depth_first_search(grid, row - 1, col, visit)
    count += depth_first_search(grid, row, col + 1, visit)
    count += dfs(grid, row, col - 1, visit)
```

**Review by ChrisO345 — APPROVED** [review]

**Pull request merged**